File: lua/test/trajtrap.lua
#!/usr/bin/env lua

---@diagnostic disable: redefined-local
package.path = arg[0]:sub(0, -arg[0]:match("([^/\\]*)$"):len() - 1) .. "?.lua;" .. package.path
local test = require("test")
local a = require("liba")
test.dir(getmetatable(a.trajtrap))
local ctx = a.trajtrap.new()
assert(ctx)
test.dir(getmetatable(ctx))
assert(a.trajtrap.gen(ctx, 2, 2, -2, 0, 4) > 0)
assert(type(a.trajtrap.pos(ctx, 0.5)) == "number")
assert(type(a.trajtrap.vel(ctx, 0.5)) == "number")
assert(type(a.trajtrap.acc(ctx, 0.5)) == "number")
assert(ctx.gen(ctx, 2, 2, -2, 0, 4) > 0)
assert(type(ctx.pos(ctx, 0.5)) == "number")
assert(type(ctx.vel(ctx, 0.5)) == "number")
assert(type(ctx.acc(ctx, 0.5)) == "number")
assert(ctx:gen(2, 2, -2, 0, 4) > 0)
assert(type(ctx:pos(0.5)) == "number")
assert(type(ctx:vel(0.5)) == "number")
assert(type(ctx:acc(0.5)) == "number")
assert(type(ctx.t) == "number")
assert(type(ctx.q0) == "number")
assert(type(ctx.q1) == "number")
assert(type(ctx.v0) == "number")
assert(type(ctx.v1) == "number")
assert(type(ctx.vc) == "number")
assert(type(ctx.ta) == "number")
assert(type(ctx.td) == "number")
assert(type(ctx.qa) == "number")
assert(type(ctx.qd) == "number")
assert(type(ctx.ac) == "number")
assert(type(ctx.de) == "number")
ctx.__name = nil
assert(ctx.__name)
ctx.__index = nil
assert(ctx.__index)
ctx.__newindex = nil
assert(ctx.__newindex)
---@class a.trajtrap
---@field __name string
---@field __index table
---@field __newindex table

File: python/test/trajtrap.py
#!/usr/bin/env python
import os, sys

path = os.path.dirname(__file__)
path = os.path.dirname(path)
sys.path.insert(0, path)
if len(sys.argv) > 1:
    sys.stdout = open(sys.argv[1], "w")

import liba  # type: ignore

traj = liba.trajtrap()
traj.gen(2, 2, -2, 0, 4)

N = 1000
T = traj.t / N
x = liba.new_num(N)
for i in range(N):
    x[i] = i * T

p = traj.pos(x)
v = traj.vel(x)
a = traj.acc(x)

for i in range(N):
    print("%g,%g,%g,%g" % (x[i], p[i], v[i], a[i]))

traj.t
traj.p0
traj.p1
traj.v0
traj.v1
traj.vc
traj.ta
traj.td
traj.pa
traj.pd
traj.ac
traj.de
